_posts/defining-messages.md

```diff
@@ -0,0 +1,169 @@
+
+Now that I said what I wanted let me try to articulate that better. In the context of distributed systems and messaging, serialization enters the stage to handle the transformation of object instances into something that can be transferred on the wire.
+
+> More details about messages in [Back to basics: Commands, events, and messages](https://milestone.topics.it/2023/05/25/back-to-basics-messages.html).
+
+For example, take into account the following JavaScript object:
+
+```javascript
+var person = {
+  "firstName": "Mauro",
+  "lastName": "Servienti"
+};
+```
+
+To transfer it on the wire as the body of a message on a queue, we need to transform it into something like this:
+
+```json
+ {
+  "firstName": "Mauro",
+  "lastName": "Servienti"
+}
+```
+
+The result in the JavaScript case is similar to the source object and can be obtained using the `JSON.stringify(person)` method.
+
+Serialization and its counterpart deserialization are pretty straightforward. We start from an object instance and get some text representation in the case of JSON. Through deserialization, we can reverse the process. From the text representation, we obtain an object instance.
+
+> There are plenty of serialization formats, all with pros and cons. They are not the topic of this article. Even if, sometimes, I feel they are also mass distraction weapons 🤪
+
+## What should we use to define messages?
+
+When dealing with message-based systems and .NET, the problem is finding the best coding structure to define messages and later serialize and transfer them on the wire.
+
+Shall we define messages as regular C# classes (aka [POCO, plain old CLR objects](https://en.wikipedia.org/wiki/Plain_old_CLR_object))? Are interfaces better? Or the recently introduced records obsolete everything else, and we should rewrite our systems?
+
+> I've witnessed people breaking friendships over those discussions. It's as hot as tabs vs spaces. It's spaces, obviously! 🤪
+
+## POCO
+
+One could say (and I'm one of them), "Keep it simple, stupid!" ([KISS](https://en.wikipedia.org/wiki/KISS_principle)) and use POCO classes. For example:
+
+```csharp
+public class OrderAccepted
+{
+   public string OrderNumber { get; set; }
+}
+```
+
+The presented class represents an event in a system dealing with orders. 
+The presented class comes with a few potential flaws that are worth listing (in no particular order):
+
+1. The `OrderAccepted` type is mutable. After deserializing the message, a receiver can change the `OrderNumber` property value. And that might even be worse because an event must be immutable (for the purists), considering it represents the past.
+2. [Contracts evolution](https://milestone.topics.it/2022/07/04/messages-evolution.html) might be tricky because using classes makes adopting a multiple inheritance approach difficult if not impossible.
```

Comment:
```suggestion
2. [Contracts evolution](https://milestone.topics.it/2022/07/04/messages-evolution.html) might be tricky because using classes makes adopting a multiple inheritance approach difficult, if not impossible.
```